auth0/resource_auth0_hook.go

```diff
@@ -96,9 +104,85 @@ func updateHook(d *schema.ResourceData, m interface{}) error {
 	if err != nil {
 		return err
 	}
+	if err = upsertHookSecrets(d, m); err != nil {
+		return err
+	}
 	return readHook(d, m)
 }
 
+func upsertHookSecrets(d *schema.ResourceData, m interface{}) error {
```

Comment:
This method is particularly tricky to comprehend. Is there a way to simplify diffing what has changed? 

Maybe if the schema type is changed (maybe to a set?) the `Diff()` method can help list what needs to be added/removed.

Reply:
You're saying instead of a Map use a Set of objects with a key and value field? Even if I did that how would "Diff" help?  That's just going to tell me the ones that changed but not the ones that got added or removed (which I'd need to know separately). I'd still end up using "setintersection" and "setsubtract" which I'm not even 100% sure works right with objects (I can find no documentation related to it).  Additionally I'm not using the current state to calculate what keys need to be added/removed, I'm calling the remote API to know and in that case I'm not going to have a Set anyway.

I dunno. I kinda prefer the map syntax to represent the concept of key value pairs. The "client_metadata" on a "client" is a Map.  While I agree the function is slightly complex I'm confused how it's difficult to comprehend. It calculates what keys need to be removed, what keys to add and what was modified and calls the appropriate API functions for each.

I have to say I absolutely hate the API to manage hook secrets.   I filed an issue with Auth0 directly about it but nothing came of it.  They could have simply made a single PUT request more flexible and stop erroring when you specify keys that didn't exist and this would have been a simple solution.

Reply:
@alexkappa about to onboard auth0 and would like to use this provider. have you had a chance to review the response to this review?

Reply:
I was thinking whether the [`Diff`](https://github.com/alexkappa/terraform-provider-auth0/blob/master/auth0/resource_data.go#L338) would help (if `secrets` were a set). It's used elsewhere in the provider and returns a list of items to add / remove.

But I see your point, the terraform configuration is better represented as a map.

That being said, this code can be simplified for readability. I've created https://github.com/go-auth0/auth0/pull/176 to help with this. Would you like to give it a review?

Reply:
Sorry for never responding.  I'm totally fine with your changes.  Would love to see this feature available through the official provider.

Reply:
I have now rebased against master and changed my code to use the new ReplaceSecrets function in auth0-go and squashed into a single commit.

This should be good now. It passes all tests:
```
$ make testacc TESTS=TestAccHookSecrets
==> Checking that code complies with gofmt requirements...
?       github.com/alexkappa/terraform-provider-auth0   [no test files]
=== RUN   TestAccHookSecrets
--- PASS: TestAccHookSecrets (6.92s)
PASS
coverage: 8.4% of statements
ok      github.com/alexkappa/terraform-provider-auth0/auth0     7.192s  coverage: 8.4% of statements
?       github.com/alexkappa/terraform-provider-auth0/auth0/internal/debug      [no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok      github.com/alexkappa/terraform-provider-auth0/auth0/internal/random     0.218s  coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok      github.com/alexkappa/terraform-provider-auth0/auth0/internal/validation 0.097s  coverage: 0.0% of statements [no tests to run]
?       github.com/alexkappa/terraform-provider-auth0/version   [no test files]
```